activestorage/app/models/active_storage/blob.rb

```diff
@@ -49,26 +50,22 @@ def find_signed(id)
       find ActiveStorage.verifier.verify(id, purpose: :blob_id)
     end
 
-    # Returns a new, unsaved blob instance after the +io+ has been uploaded to the service.
-    # When providing a content type, pass <tt>identify: false</tt> to bypass automatic content type inference.
-    def build_after_upload(io:, filename:, content_type: nil, metadata: nil, identify: true)
-      new(filename: filename, content_type: content_type, metadata: metadata).tap do |blob|
-        blob.upload(io, identify: identify)
-      end
-    end
-
     def build_after_unfurling(io:, filename:, content_type: nil, metadata: nil, identify: true) #:nodoc:
       new(filename: filename, content_type: content_type, metadata: metadata).tap do |blob|
         blob.unfurl(io, identify: identify)
       end
     end
 
-    # Returns a saved blob instance after the +io+ has been uploaded to the service. Note, the blob is first built,
-    # then the +io+ is uploaded, then the blob is saved. This is done this way to avoid uploading (which may take
-    # time), while having an open database transaction.
-    # When providing a content type, pass <tt>identify: false</tt> to bypass automatic content type inference.
-    def create_after_upload!(io:, filename:, content_type: nil, metadata: nil, identify: true)
-      build_after_upload(io: io, filename: filename, content_type: content_type, metadata: metadata, identify: identify).tap(&:save!)
+    # Creates a new blob instance and then uploads the contents of
+    # the given <tt>io</tt> to the service. The blob instance is going to
+    # be saved before the upload begins to prevent the upload clobbering another due to key collisions.
+    # When providing a content type, pass <tt>identify: false</tt> to bypass
+    # automatic content type inference.
+    def create_and_upload!(io:, filename:, content_type: nil, metadata: nil, identify: true)
+      build_after_unfurling(io: io, filename: filename, content_type: content_type, metadata: metadata, identify: identify).tap do |blob|
+        blob.save! # To ensure the uniqueness constraint in the database fires before uploading
+        blob.upload(io, identify: identify)
```

Comment:
Note that we want to `upload_without_unfurling`, since we’ve already computed the blob’s checksum and extracted its content type in `build_after_unfurling`.